### PR TITLE
Adds questionnaire endpoints for end user job editing

### LINKED ISSUE #14: Provide API endpoints for Job Questionnaire
CWL has an input json/yaml file that contains parameters for running the job.
Some of these parameters will be hard coded by the system and some will be user editable.
Each parameter is like a question that needs to be answered.

So for each WorkflowVersion(CWL file) we will have JobQuestionnaire that contains a list of JobQuestions(for each parameter). In order to build the CWL input file all JobQuestions will need one or more JobAnswer.

__Job Creation Workflow__
User will pick a WorkflowVersion(RnaSeq for example).
User lists JobQuestionnaires for that WorkflowVersion and gets a list:
- RnaSeq for Ant
- RnaSeq for Human

User creates a JobAnswerSet for the `RnaSeq for Ant` JobQuestionnaire
User lists JobQuestions for that JobAnswerSet/JobQuestionnaire.
User writes JobAnswer for a user editable question.

User requests the JobAnswerSet to be converted into a Job.
Job fields including input json are filled in from the JobQuestionnaire.
JobInputFile objects are created based on the JobQuestionnaire

### PR DESCRIPTION
**Pull request by johnbradley**
Adds endpoints:

- `/api/job-questions/` - contains all questions that could be answered for a job
- `/api/job-questionnaires/` - questionnaire is a list of questions required for a particular job
- `/api/job-answers/` - answers to a question - if tied to a questionnaire they are read only - otherwise editable by the user who created it
- `/api/job-string-answers/` - string value associated with an answer
- `/api/job-dds-file-answers/` - DukeDS file info associated with an answer
- `/api/job-answer-sets/` - group of user answers and the questionnaire they apply to

Fixes #14

### CODE REVIEW

File: data/models.py [hunk 1]
```diff
@@ -191,3 +191,123 @@ class LandoConnection(models.Model):
...
+
+class JobStringAnswer(models.Model):
+    """
+    String value associated with a JobAnswer.
+    """
+    answer = models.OneToOneField(JobAnswer, on_delete=models.CASCADE, primary_key=True, related_name='string_value')
+    value = models.CharField(max_length=255, blank=False, null=False)
+
+    def __unicode__(self):
+        return '{} question:{} - value:{}'.format(self.id, self.answer.question.name, self.value)
```
Comment (dleehr) [inline data/models.py]: By setting `primary_key=True`, the `id` field is not added, so this raises an exception when I try to create a JobStringAnswer through the Admin UI. `self.pk`should work.

> AttributeError at /admin/data/jobstringanswer/add/
> 'JobStringAnswer' object has no attribute 'id'

https://docs.djangoproject.com/en/1.10/ref/models/fields/#primary-key

File: data/models.py [hunk 2]
```diff
@@ -191,3 +191,123 @@ class LandoConnection(models.Model):
...
+    answer = models.OneToOneField(JobAnswer, on_delete=models.CASCADE, primary_key=True, related_name='dds_file')
+    project_id = models.CharField(max_length=255, blank=False, null=True,
+                                  help_text='uuid from DukeDS for the project containing our file')
+    file_id = models.CharField(max_length=255, blank=False, null=True,
+                               help_text='uuid from DukeDS for the file chosen as this answer')
+    dds_user_credentials = models.ForeignKey(DDSUserCredential, on_delete=models.CASCADE,
+                                             help_text='Credentials with access to this file')
+
+    def __unicode__(self):
+        return '{} question:{} - file_id:{}'.format(self.id, self.answer.question.name, self.file_id)
```
Comment (dleehr) [inline data/models.py]: Same as above. setting a custom `primary_key` field means that `self.id` doesn't exist.

**Review by dleehr — CHANGES_REQUESTED** [review]
I was able to run the migrations and

1. add a number of questions
2. add a questionnaire

Adding answers failed because of the above mentioned issue, so this should be addressed.

Also, if a system answer is to be a file path (e.g. a reference genome that will be on the VM), would that be a string answer? What about an integer (like number of threads?)

**Comment by johnbradley** [comment]
> Also, if a system answer is to be a file path (e.g. a reference genome that will be on the VM), would that be a string answer? 

Yes. My thought was the question would be data_type `File` and the answer would be data type `string`

>  What about an integer (like number of threads?)

For right now, yes. My thought was the question would be data_type `int` and the answer would be data type `string` So we convert the string to an integer before building CWL input json. I can add __job-integer-answers__ if it would help.

There has been no code written(yet) that converts the questionnaire to a CWL input JSON blob.

**Comment by dleehr** [comment]
That's fine. We can work with strings, just wanted to make sure it would be accepted. 

I'm testing this locally with Ember. As it exists now, Ember will require special-casing for the endpoints that don't have an `id` field (the ones where primary_key is overridden). Is there a big benefit to breaking this convention on the Django side? I gather that we'd add a field in our table, but it bucks the convention and requires special-casing on the client.

**Comment by johnbradley** [comment]
>I'm testing this locally with Ember. As it exists now, Ember will require special-casing for the endpoints that don't have an id field (the ones where primary_key is overridden). Is there a big benefit to breaking this convention on the Django side? I gather that we'd add a field in our table, but it bucks the convention and requires special-casing on the client.

I was just following the Django example for OneToOneField (https://docs.djangoproject.com/en/1.10/topics/db/examples/one_to_one/).
I can remove the primary_key property and see what happens.

**Comment by dleehr** [comment]
I can try that here, was just curious of the reasoning. It makes sense on the django side, since the field would be unique anyways and is required. I suppose we could even just add an `id` field to the serializer, as a copy of the answer field, but I'll investigate what Ember needs.

**Comment by johnbradley** [comment]
I removed the `primary_key=True` property from those two tables. Everything seems to still work fine. I'm fine with this change if it helps the front-end and back-end fit together better.

**Comment by johnbradley** [comment]
@dleehr Do you have further changes requested that I am missing?

**Review by dleehr — APPROVED** [review]

**Pull request merged**